### ISSUE TITLE
ComicVIne has blocked ComicTagger

### ISSUE THREAD

**Issue by eduo** [comictagger/comictagger#501]
(I apologize. I saw an older issue that seemed to deal with the comicvine API rate limiting and since it was not closed I assumeed it was still under revision. My mistake)

Is there a way to look at logging related to the Comicvine API? 

I'm on mac and checked both the built-in log window and the console but I can't find the specific of the error I've been getting the past two days: Searching for metadata fails and testing the API says the API is not valid.

<img width="532" alt="image" src="https://github.com/comictagger/comictagger/assets/154630/c61e8ce7-10e5-4655-b846-1808c437ef2a">

I can access Comicvine OK, and I can get results querying the API via direct URL with my API Key. Removing the API key from Comicvine doesn't make a difference in the messages.

https://comicvine.gamespot.com/api/volumes/?api_key=XXXXX_APIKEY&format=json&sort=name:asc&filter=name:Walking%20Dead

I've been doing a ton of searches lately, refreshing my library's tags (mostly migrating to comicrack and getting the missing information), but for the past two days it hasn't worked. I thought my IP or my API Key were borked but trying manually doesn't seem to give problems.

When seeing the API details it says it has no limits.

![image](https://github.com/comictagger/comictagger/assets/154630/1983f1c1-c389-4e75-b9dd-495385e68c66)

**Comment by Neodemonhawk** [comment]
Same issue with v1.5.5 on Windows. Can access comicvine manually fine and even scraping in comicrack with scraper script works with the same api key.

**Comment by lordwelch** [comment]
So it appears that comicvine has specifically blocked ComicTagger: https://comicvine.gamespot.com/forums/api-developers-2334/comictagger-blocked-on-comicvine-2316231/

You can try this and see that it fails
`curl -i 'https://comicvine.gamespot.com/api/search/?api_key=63a92cc70ce2d0b371a73c1071cffc8921243d95&forma
t=json&resources=volume&query=test&field_list=volume%2Cname&page=1&limit=1' -H 'User-Agent: comictagger/1.5.5'`

You can replace comictagger with literally anything else and then it will work
`curl -i 'https://comicvine.gamespot.com/api/search/?api_key=63a92cc70ce2d0b371a73c1071cffc8921243d95&forma
t=json&resources=volume&query=test&field_list=volume%2Cname&page=1&limit=1' -H 'User-Agent: literally anything else/1.5.5'`

**Comment by dquillen312** [comment]
I thought I was going crazy. Can confirm as well... replaced the user-agent in postman with comictagger/1.5.5 and no go. Put it back to PostmanRuntime/7.32.3 and worked without issue.

**Comment by eduo** [comment]
Interesting. I guess it's time to put the user agent string as a configurable field like many browsers do :(

**Comment by bpepple** [comment]
> Interesting. I guess it's time to put the user agent string as a configurable field like many browsers do :(

That's a really bad idea. Clearly, ComicVine has an issue with ComicTagger and circumventing their block is a good way to give ComicVine a reason to shut down their API (which is a **free** service that as far as I'm aware isn't monetized and costs them money).

It's worth contacting ComicVine to see if there is something the devs can modify in CT to alleviate ComicVine's concern. Now, getting in contact with ComicVine will most likely take awhile since they don't seem to monitor the dev forums very closely.

**Comment by beville** [comment]
It's unfortunate.  They definitely should have reached out first.  Certainly in the past no one there ever made a stink about using their publicly available API as long as every user had a key and respected the rate limit. 

I suspect that  some manager at the new owners was seeing that the most common user agent string  that was hitting the API had "comictagger" in it, and decided to save some bit of bandwidth.  If that's the case, the issue isn't that ComicTagger in particular is abusing the system, but it might be most _identifiable_ application hitting the API.  I do note at least one other popular open source app that uses a random user agent for each CV connection via [user_agent2](https://github.com/dytttf/user_agent2).  

The Terms of Use [here](https://comicvine.gamespot.com/api/#:~:text=Terms%20of%20Use,thousands%20of%20contributors.) don't appear to have changed.  I don't think ComicTagger's usage patterns have changed.  It may be that its overall usage has gone up, though.

But if they're making moves like this then I'm guessing the API is already on the edge of being shut down.

**Comment by bpepple** [comment]
> I suspect that some manager at the new owners was seeing that the most common user agent string that was hitting the API had "comictagger" in it, and decided to save some bit of bandwidth. If that's the case, the issue isn't that ComicTagger in particular is abusing the system, but it might be most _identifiable_ application hitting the API. I do note at least one other popular open source app that uses a random user agent for each CV connection via [user_agent2](https://github.com/dytttf/user_agent2).

Yeah, someone told me Mylar was doing that. I'm not a fan of trying to gaming the system like that, since it makes debugging problems tougher for the server admins and gives ComicVine more incentive to close down the API.
 
> The Terms of Use [here](https://comicvine.gamespot.com/api/#:~:text=Terms%20of%20Use,thousands%20of%20contributors.) don't appear to have changed. I don't think ComicTagger's usage patterns have changed. It may be that its overall usage has gone up, though.

Agreed. I'm guessing CT's usage has gone up a decent amount over the years, and all those new users probably just shit-hammer the API tagging their whole collections when they first start using it. 😱

> But if they're making moves like this then I'm guessing the API is already on the edge of being shut down.

Yup, the writings been on the wall for awhile now, and I wouldn't be surprised they stop offering it in the next year or two.

**Comment by lordwelch** [comment]
The next version of CT implements some proper rate limiting (and some severe rate limiting for the builtin key). But even just the standard speed of CT didn't go much over. I also have plugins for CT at like 90% completed so when you are pip installing CT you can just add another package (currently the only one that exists is https://github.com/mizaki/comictagger-metron-talker) and it will show up in the GUI.

I'm also not a fan of trying to get around the system, especially when there are other options (Marvel API, Metron, using a GCD export, various manga websites)

**Comment by lordwelch** [comment]
I have opened a ticket on their ZenDesk, we'll see what happens

**Comment by mizaki** [comment]
> The next version of CT implements some proper rate limiting (and some severe rate limiting for the builtin key). But even just the standard speed of CT didn't go much over. I also have plugins for CT at like 90% completed so when you are pip installing CT you can just add another package (currently the only one that exists is https://github.com/mizaki/comictagger-metron-talker) and it will show up in the GUI.
> 
> I'm also not a fan of trying to get around the system, especially when there are other options (Marvel API, Metron, using a GCD export, various manga websites)

I do have one for GCD too but it involves converting the mysql dump to sqlite with a random script I found. (I looked for a python sqldump to sqlite but couldn't find one.) I will ask GCD if they are willing to do a sqlite export direct. (Or someone else can if they want to get there first.) If people want to test out the alpha version, I can put it up.

The Metron one I'm not sure if anyone other than me has tested but should be working.

**Comment by eduo** [comment]
> > The next version of CT implements some proper rate limiting (and some severe rate limiting for the builtin key). But even just the standard speed of CT didn't go much over. I also have plugins for CT at like 90% completed so when you are pip installing CT you can just add another package (currently the only one that exists is https://github.com/mizaki/comictagger-metron-talker) and it will show up in the GUI.
> > I'm also not a fan of trying to get around the system, especially when there are other options (Marvel API, Metron, using a GCD export, various manga websites)
> 
> I do have one for GCD too but it involves converting the mysql dump to sqlite with a random script I found. (I looked for a python sqldump to sqlite but couldn't find one.) I will ask GCD if they are willing to do a sqlite export direct. (Or someone else can if they want to get there first.) If people want to test out the alpha version, I can put it up.
> 
> The Metron one I'm not sure if anyone other than me has tested but should be working.

I was not aware GCD made the SQL database available. I can't find this in their resource pages. Do you have a link?

**Comment by bpepple** [comment]
> I do have one for GCD too but it involves converting the mysql dump to sqlite with a random script I found. (I looked for a python sqldump to sqlite but couldn't find one.) I will ask GCD if they are willing to do a sqlite export direct. (Or someone else can if they want to get there first.) If people want to test out the alpha version, I can put it up.

I'd find it unlikely GCD will handle the conversion, since Django doesn't natively have the option to dump in a different database format, tho they could dump it in json or yml format.

**Comment by DieselTech** [comment]
> I was not aware GCD made the SQL database available. I can't find this in their resource pages. Do you have a link?

https://docs.comics.org/wiki/Main_Page#Database

> I'd find it unlikely GCD will handle the conversion, since Django doesn't natively have the option to dump in a different database format, tho they could dump it in json or yml format.

I'm not dev, but are you sure on that? It seems from their documentation they do support it: https://docs.djangoproject.com/en/4.2/ref/databases/#sqlite-notes

If we change line 102 here to "django.db.backends.sqlite" we should be able to create a base install using sqlite. 

https://github.com/GrandComicsDatabase/gcd-django/blob/cebb581bfe9f10e3d239f67d8385fb76ce8170d2/settings.py#L102C29-L102C29

**Comment by bpepple** [comment]
> I'm not dev, but are you sure on that? It seems from their documentation they do support it: https://docs.djangoproject.com/en/4.2/ref/databases/#sqlite-notes
> 
> If we change line 102 here to "django.db.backends.sqlite" we should be able to create a base install using sqlite.
> 
> https://github.com/GrandComicsDatabase/gcd-django/blob/cebb581bfe9f10e3d239f67d8385fb76ce8170d2/settings.py#L102C29-L102C29

@DieselTech Umm, that would require the GCD to change what DB they're running in **production**, which I think I can confidently say they won't do. It's not a trivial thing to switch DB back-ends on a long running Django project, since you would need handle the historical model migrations and then export the existing MySQL data (about 2GB) into SQLite db (and that's assuming they aren't using some MySQL-only features).

**Comment by DieselTech** [comment]
> that would require the GCD to change what DB they're running in production

Oh no I wasn't suggesting that GCD make the changes. I'm not expecting them to do anything more than just provide the .sql dump. I was saying WE could fork the project and make the changes needed. It's going to be massive effort on our end but I'm willing to help with it if it means the outcome is a sqlite database.

**Comment by bpepple** [comment]
@DieselTech Oh, if that's the case it's easier to run something like ***mysql2sqlite*** to convert it to a SQLite dump, than bother setting up a Django instance to accomplish the same thing.

This tangent came up since @mizaki mentioned contacting GCD about providing a SQLite dump, and I didn't really think that was something that'd be easy for them to do.

**Comment by DieselTech** [comment]
I looked around for months for something that did just that, even considered writing my own. Damn that saves me time. Well I have it in sqlite format and ran a few checks on it to make sure everything looks good so far. 

If needed I can re-host the file for others to grab. It ballooned to 4.5GB

**Comment by IDontKnow218** [comment]
Also having the same problem. Any temporary fix for this?? @lordwelch

**Comment by lordwelch** [comment]
I'm not going to support any workarounds for ComicTagger, CV has specifically chosen to block it and that is their choice.

Now comicvine scraper I don't believe is blocked
https://github.com/cbanack/comic-vine-scraper it is a plugin for ComicRack which you can download from the internet archive https://web.archive.org/web/20181018173957/http://comicrack.cyolito.com/downloads
You can also use Mylar https://github.com/mylar3/mylar3 they use an old modified version of CT internally that still works due to their modifications.

In the next few weeks I plan on having at least one new source as an option for ComicTagger so it will no longer depend on the good graces of ComicVine to be useful.

Currently still no response from ComicVine on why CT was blocked, Fandom is a fairly small company with a lot of websites so I'm not surprised that they haven't responded yet.

**Comment by carterartist** [comment]
So, why is comicvine supposedly blocking comictagger, yet not reflecting this with the API code page?

**Comment by lordwelch** [comment]
We don't know why. But how a block is implemented is almost never documented it's usually expected that you'll notice if your blocked. In this case if you use a user-agent identifying the app as ComicTagger you get an [http 403] error code which means forbidden 

[http 403]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/403

**Comment by protofolius** [comment]
Simply start sending out a generic useragent string .
The same thing has happened to other another project too , 1 year ago . And that is what they did to work around it .

**Comment by lordwelch** [comment]
@protofolius See https://github.com/comictagger/comictagger/issues/501#issuecomment-1692066852 and https://github.com/comictagger/comictagger/issues/501#issuecomment-1694736329 for why I won't be doing this. If ComicVine doesn't want us to use their service then I won't. [Metron] was started specifically to make an alternative to ComicVine and is a viable alternative among others. If ComicVine does not budge on blocking ComicTagger, I will move the ComicVine code to a separate repository and those that wish to use a workaround will be able to make their modifications on a fork. I have plans for loading plugins from the ComicTagger settings folder to make ComicTagger a little more useful even when it goes into disrepair (there are still plugins for ComicRack that get use today even though it's been years since it got an update, and you have to go to the Internet Archive to even download it anymore). However, there are currently a lot of changes since the last release [1.5.5] and some things are still in-flux or need further refinement before I work on that in particular. The next alpha release will have some other source as an option.

[Metron]: https://github.com/bpepple/metron#-motivation
[1.5.5]: https://github.com/comictagger/comictagger/releases/tag/1.5.5

**Comment by lordwelch** [comment]
I'm getting movement on my ticket! He's pinging the engineering team. We'll see what happens.

**Comment by DieselTech** [comment]
> I'm getting movement on my ticket! He's pinging the engineering team. We'll see what happens.

Are the ticket views public or are they locked down to your account?

**Comment by lordwelch** [comment]
They are limited to the account, I think if I cc you on the email you would be able to see it. My CV account is using advert@narnian.us Send me an email and I can try.

**Comment by tacotruffles** [comment]
ComicVine API is responding now!

**Comment by lordwelch** [comment]
I have an email where someone said it's working now, and sure enough it is, but their comment appears to be lost in the void...

**Comment by mrnejc** [comment]
I can confirm it works with 1.5.5 available thru pip3 install and v1.6.0a5.dev70 in development branch:

![Screenshot from 2023-08-31 11-41-47](https://github.com/comictagger/comictagger/assets/1905477/2c0910b7-3fcb-4d79-bda7-19bd257399a7)

**Comment by mizaki** [comment]
Hopefully they come back with why it was done and how it won't be done again. 

It does give a little renewed imputes to the support for other sources so if something does happen to CV, it's not a total show stopper.

**Comment by lordwelch** [comment]
I closed out the ticket yesterday, they unblocked it but didn't provide a reason. I'm just happy they unblocked us. I'll leave this issue open until the next alpha release.

**Comment by lordwelch** [comment]
ComicVIne is unblocked and [1.6.0-alpha.6: Hal Jordan] now has https://metron.cloud as an alternative. This issue is now resolved.

[1.6.0-alpha.6: Hal Jordan]: https://github.com/comictagger/comictagger/releases/tag/1.6.0-alpha.6
